fix(coreDirectives): Added an Error Message for invalid selectors

danrasmuson opened the pull request:
if the component selector is an empty string, not a string, or not supplied an error will be thrown

Closes #3464

danielrasmuson commented:
@pkozlowski-opensource here is a simple start. Let me know how I should modify it / add to it. Excited to help.

vicb commented on modules/angular2/src/core/metadata/directives.ts:
```diff
@@ -859,6 +860,9 @@ export class ComponentMetadata extends DirectiveMetadata {
     viewBindings?: List<any>,
     changeDetection?: string,
   } = {}) {
+    if (!isString(selector) || !selector || !selector.length) {
```
`!selector` would not work with dart, you should use `isBlank(selector)`

mhevery commented:
@danielrasmuson can you clean up the PR pr @vicb  comments?

danielrasmuson commented:
@mhevery Thanks for tag.

@pkozlowski-opensource I've updated this PR to not throw when the selector is not provided (#3464). So currently it only throws if you provide a non-string or a string of length 0.

At the time of this writing, if you don't provide a selector 
It will render like this.
```html
<undefined class="ng-binding"><h1>Hello</h1></undefined>
```
I'm assuming this will change once #1662 is completed.

Lastly, I'm noticing I'm having some dart problems. How can I fix those?

vicb commented on modules/angular2/test/core/compiler/element_injector_spec.ts:
```diff
@@ -513,6 +513,25 @@ export function main() {
+
+      it('should throw an error if selector is not a string', () => {
+        expect(function(){
+          new ComponentMetadata({selector: {x: 'selector'}});
```
The changes LGTM however this test triggers an error (TS, Dart) because the expected type is string. May be you should remove it ?

danielrasmuson replied:
Thanks @vicb, I took your advice and removed that scenario.

danrasmuson committed:
fix(coreDirectives): Added an Error Message for invalid selectors on component annotation

If the component selector is an empty string or not a string an error will be thrown.

Closes #3464

danielrasmuson commented:
I'm getting the below dart error. Should I put this into another file?

```
TS2DartError: [TSToDartTranspiler]: angular2/src/core/metadata/directives.ts:834:5: const constructors can only contain assignments and super calls
```

danielrasmuson commented:
Thanks @vicb I implemented your recommendations. Let me know your thoughts.

tbosch commented on modules/angular2/src/core/render/dom/compiler/selector.ts:
```diff
@@ -32,6 +34,10 @@ export class CssSelector {
 
   static parse(selector: string): CssSelector[] {
+    if (!isBlank(selector) && !(isString(selector) && selector.length > 0)) {
+      throw makeTypeError('The component annotation requires a string selector.');
```
Could you throw a `BaseException`, and use the following error:
`Empty selectors are not allowed`

danielrasmuson commented:
Thanks for the feedback @tbosch. I've implemented your changes.

vicb commented on modules/angular2/src/core/render/dom/compiler/selector.ts:
```diff
@@ -32,6 +33,10 @@ export class CssSelector {
   notSelectors: CssSelector[] = [];
 
   static parse(selector: string): CssSelector[] {
+    if (!isBlank(selector) && !(isString(selector) && selector.length > 0)) {
```
1) please rewrite as 
```js
if (StringWrapper.equals(selector, '')) {
  throw new BaseException('Selectors must not be an empty string');
}
```

2) Update the test with the new error message (and please use single quotes consistently)

3) rebase

And we'll merge asap (sorry for the delay)

danielrasmuson commented:
@vicb Thanks Victor for the comments. I made the changes you suggested, however the tests are failing, my formatter might be outdated. I'll look into it when a get a moment and get back to you.

vicb commented:
@danielrasmuson I had updated the code in the linked PR above (4959). There are remaining issues with Dart. I want to chat with @tbosch to find the best possible solution.

vicb commented:
@danielrasmuson after having discussed with Tobias we think that this PR as is as little value:
- The check should not be in the selector but in the compiler,
- It will only catch the case the user explicitly set the selector to an empty string.

I'm closing for now and I'll comment on the original issue.

Thanks for your work on the PR and sorry for the delay in processing.

danielrasmuson commented:
Originally this pr erred on any selector that wasn't a string, which provides a lot more value. However, as other opinions came to light it transformed into its current state. @vicb I agree asserting the selector isn't an empty string isn't very valuable. Thanks for your feedback and I'll check out another issue.
